src/cocotb/_scheduler.py

```diff
@@ -341,170 +368,121 @@ def _test_completed(self, trigger=None):
-                        )
+        Finds all Tasks waiting on the Trigger that fired and queues them.
+        """
+        if _debug:
+            self.log.debug(f"Trigger fired: {trigger}")
+
+        # find all tasks waiting on trigger that fired
+        try:
+            scheduling = self._trigger2tasks.pop(trigger)
+        except KeyError:
+            # GPI triggers should only be ever pending if there is an
+            # associated task waiting on that trigger, otherwise it would
+            # have been unprimed already
+            if isinstance(trigger, GPITrigger):
+                self.log.critical(f"No tasks waiting on trigger that fired: {trigger}")
+                trigger.log.info("I'm the culprit")
+            # For Python triggers this isn't actually an error - we might do
+            # event.set() without knowing whether any tasks are actually
+            # waiting on this event, for example
+            elif _debug:
+                self.log.debug(f"No tasks waiting on trigger that fired: {trigger}")
+            return
 
-                    del trigger
-                    continue
+        if _debug:
+            debugstr = "\n\t".join([str(task) for task in scheduling])
+            if len(scheduling) > 0:
+                debugstr = "\n\t" + debugstr
+            self.log.debug(
+                f"{len(scheduling)} pending tasks for trigger {trigger}{debugstr}"
+            )
 
-                if _debug:
-                    debugstr = "\n\t".join([str(task) for task in self._scheduling])
-                    if len(self._scheduling) > 0:
-                        debugstr = "\n\t" + debugstr
-                    self.log.debug(
-                        f"{len(self._scheduling)} pending tasks for trigger {trigger}{debugstr}"
-                    )
+        # queue all tasks to wake up
+        for task in scheduling:
+            # unset trigger
+            task._trigger = None
+            self._queue(task, outcome=trigger._outcome)
 
-                # This trigger isn't needed any more
-                trigger._unprime()
+        # This trigger isn't needed any more
+        trigger._unprime()
 
-                for task in self._scheduling:
-                    if task._outcome is not None:
-                        # Task was killed by another task waiting on the same trigger
-                        continue
-                    if _debug:
-                        self.log.debug(f"Scheduling task {task}")
-                    self._schedule(task, trigger=trigger)
-                    if _debug:
-                        self.log.debug(f"Scheduled task {task}")
+    def _event_loop(self) -> None:
+        """Run the main event loop.
 
-                    # remove our reference to the objects at the end of each loop,
-                    # to try and avoid them being destroyed at a weird time (as
-                    # happened in gh-957)
-                    del task
+        This should be triggered by only:
```

Comment:
This phrasing sounds more natural to me.
```suggestion
        This should only be started by:
```